### PR TITLE
[1/n][vllm, rollout] feat: flowgrpo - support vllm-omni as rollout backend for verl

### PR DESCRIPTION
**Pull request by knlnguyen1802**
### What does this PR do?
Follow up the work from https://github.com/verl-project/verl/pull/5297
This PR adds **vLLM-Omni** as a rollout backend, enabling diffusion-based image generation models (e.g., QwenImage) in verl RLHF pipelines with log-prob extraction and LoRA weight sync.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

Unit tests added for vllm_omni rollout 
`tests/workers/rollout/rollout_vllm/test_vllm_omni_generate.py`

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

- **`vllm_base_async_server.py`** — Extracted `BaseVLLMHttpServer` / `BaseVLLMReplica` shared between vLLM and vLLM-Omni to eliminate duplication
- **`vllm_omni_async_server.py`** — `vLLMOmniHttpServer` wraps `AsyncOmni`; supports token-in/image-out generation, log-prob + latent capture, custom diffusion pipelines via config
- **`vllm_omni_rollout.py`** — `vLLMOmniServerAdapter` client adapter; lazily acquires named Ray actor and calls `collective_rpc` for weight sync
- **`verl/models/diffuser_model/vllm_omni/`** — New utility package:
  - `OmniTensorLoRARequest` + `VLLMOmniHijack` — patches `DiffusionLoRAManager._load_adapter` to accept in-memory LoRA tensors (vLLM-Omni only supports file-path loading natively)
  - `QwenImagePipelineWithLogProb` — extends `QwenImagePipeline` with per-step log probs and intermediate latents for PPO/GRPO
- **`workers/config/rollout.py`** — Consolidated `DiffusionRolloutConfig` / `DiffusionSamplingConfig` into main rollout config
- **`vllm_rollout/utils.py`** — Added `vLLMOmniColocateWorkerExtension` for colocated weight-update support

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/workers/rollout/vllm_rollout/vllm_omni_rollout.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+from verl.utils.device import get_device_id
+from verl.workers.config import HFModelConfig, RolloutConfig
+from verl.workers.rollout.vllm_rollout.utils import get_device_uuid
+from verl.workers.rollout.vllm_rollout.vllm_rollout import ServerAdapter
+
+logger = logging.getLogger(__file__)
+logger.setLevel(os.getenv("VERL_LOGGING_LEVEL", "INFO"))
+
+
+class vLLMOmniServerAdapter(ServerAdapter):
```
Comment (wuxibin89) [inline verl/workers/rollout/vllm_rollout/vllm_omni_rollout.py]: Why we need a separate `vLLMOmniServerAdapter`?

File: verl/workers/config/rollout.py [hunk 2]
```diff
@@ -291,3 +302,28 @@ def __post_init__(self):
                 raise NotImplementedError(
                     f"Current rollout {self.name=} not implemented pipeline_model_parallel_size > 1 yet."
                 )
+
+
+@dataclass
+class DiffusionRolloutConfig(RolloutConfig):
+    _mutable_fields = {"max_model_len", "load_format"}
+
+    val_kwargs: DiffusionSamplingConfig = field(default_factory=DiffusionSamplingConfig)
```
Comment (wuxibin89) [inline verl/workers/config/rollout.py]: Naming: why is `val_kwargs`?

File: verl/workers/rollout/vllm_rollout/vllm_async_server.py [hunk 3]
```diff
@@ -69,47 +61,28 @@
...
-    from vllm.utils import FlexibleArgumentParser
+    pass
 
 
 logger = logging.getLogger(__file__)
 logger.setLevel(logging.INFO)
 
 
-class vLLMHttpServer:
+class vLLMHttpServer(BaseVLLMHttpServer):
```
Comment (wuxibin89) [inline verl/workers/rollout/vllm_rollout/vllm_async_server.py]: vLLMHttpServer is used by many users and production system. It's not very intuitive after inheriting BaseVLLMHttpServer. 

Is it possible that `vLLMOmniHttpServer` inherit `vLLMHttpServer`, keep vLLMHttpServer as it is and override what vllm-omni need: server_args, generate?

File: examples/vllm_omni/pipeline_qwenimage.py [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+
+def _maybe_to_cpu(v):
+    if isinstance(v, torch.Tensor):
+        return v.detach().cpu()
+    return v
+
+
+# Custom pipeline class for QwenImage that returns log probabilities during the diffusion process.
+# This is compatible with API of vllm-omni
+class QwenImagePipelineWithLogProb(QwenImagePipeline):
```
Comment (wuxibin89) [inline examples/vllm_omni/pipeline_qwenimage.py:39]: Why not use QwenImagePipeline from diffusers? We're not going to maintain model definition in verl and reuse transformers and diffusers.

File: verl/models/diffusion/vllm_omni/qwen_image/qwen_image_transformer.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    QwenEmbedRope,
+    QwenImageTransformer2DModel,
+    QwenImageTransformerBlock,
+    QwenTimestepProjEmbeddings,
+)
+
+
+# This class is need because there is bug in vllm-omni
+# TODO: Remove this class after the bug is fixed and vllm-omni is updated to the fixed version
+class QwenImageTransformer2DModelFixed(QwenImageTransformer2DModel):
```
Comment (wuxibin89) [inline verl/models/diffusion/vllm_omni/qwen_image/qwen_image_transformer.py]: When will this be fixed?

File: examples/vllm_omni/pipeline_qwenimage.py (hunk 4 above)
Comment (knlnguyen1802) [reply examples/vllm_omni/pipeline_qwenimage.py:39]: The official pipeline does not support the functionality of stochastic diffusion process, return arbitrary returns for reinforcement learning. For example in paper like FlowGRPO/DanceGRPo we need a self-defined pipeline.
I think we can move it as an example ? Do you have any suggestion where to put it ?

File: verl/models/diffusion/vllm_omni/qwen_image/qwen_image_transformer.py (hunk 5 above)
Comment (knlnguyen1802) [reply verl/models/diffusion/vllm_omni/qwen_image/qwen_image_transformer.py]: There is some plan for fixing this as this is general problem now. Will update it

File: verl/workers/rollout/vllm_rollout/vllm_omni_rollout.py (hunk 1 above)
Comment (knlnguyen1802) [reply verl/workers/rollout/vllm_rollout/vllm_omni_rollout.py]: Found to be redundant, let's me fix this

File: verl/workers/rollout/vllm_rollout/vllm_omni_async_server.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        num_preempted = None
+        if final_res.request_output is not None and hasattr(final_res.request_output, "num_preempted"):
+            num_preempted = final_res.request_output.num_preempted
+
+        return ImageOutput(
+            image=image,
+            log_probs=log_probs,
+            stop_reason=stop_reason,
+            num_preempted=num_preempted,
+            extra_info=extra_info,
```
Comment (AndyZhou952) [inline verl/workers/rollout/vllm_rollout/vllm_omni_async_server.py]: the return should be `extra_fields`
Also better rename to `extra_fields` above for consistency

File: verl/workers/rollout/vllm_rollout/vllm_async_server.py (hunk 3 above)
Comment (knlnguyen1802) [reply verl/workers/rollout/vllm_rollout/vllm_async_server.py]: I've refactor `vLLMOmniHttpServer ` to inherit from `vLLMHttpServer`. There is some feature that need to custom for `vllm-omni` so I need to refactor out some function from `vLLMHttpServer`. But there is no actually code change

File: verl/workers/rollout/replica.py [hunk 7]
```diff
@@ -51,6 +51,19 @@ class TokenOutput(BaseModel):
     """Extra fields for dynamic addition."""
 
 
+class ImageOutput(BaseModel):
```
Comment (wuxibin89) [inline verl/workers/rollout/replica.py]: We should also support video output, e.g wan2.2

File: examples/vllm_omni/pipeline_qwenimage.py (hunk 4 above)
Comment (wuxibin89) [reply examples/vllm_omni/pipeline_qwenimage.py:39]: Is there a uniform way to monkey patch all diffusion model? If not, then we have to maintain each model definition in verl, which is a bit verbose.

File: examples/vllm_omni/scheduling_flow_match_sde_discrete.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+                variance_noise = randn_tensor(
+                    model_output.shape,
+                    generator=generator,
+                    device=model_output.device,
+                    dtype=model_output.dtype,
+                )
+                prev_sample = prev_sample_mean + std_dev_t * variance_noise
+
+            if return_logprobs:
+                log_prob = -((prev_sample.detach() - prev_sample_mean) ** 2)
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/vllm_omni/scheduling_flow_match_sde_discrete.py:223]: we follow the implementation of cps algo in FlowGRPO repo

File: verl/models/diffusion/vllm_omni/qwen_image/qwen_image_transformer.py (hunk 5 above)
Comment (zhtmike) [reply verl/models/diffusion/vllm_omni/qwen_image/qwen_image_transformer.py]: It should be fixed in vllm-omni main branch, let us drop it

File: examples/vllm_omni/pipeline_qwenimage.py (hunk 4 above)
Comment (zhtmike) [reply examples/vllm_omni/pipeline_qwenimage.py:39]: Since the FlowGRPO algorithm works based on the reverse sampling process of each model, and this process differs significantly across models, I personally think there is currently no universal way to apply a monkey patch on the vllm-omni side—unless we first modularize and split the pipeline in vllm-omni. Any suggestions? @knlnguyen1802 @SamitHuang

Comment (wuxibin89) [reply examples/vllm_omni/pipeline_qwenimage.py:39]: If vllm-omni can't modularize pipeline in short term, please move it to example folder.

Comment (SamitHuang) [reply examples/vllm_omni/pipeline_qwenimage.py:39]: agree. we will add an RFC to discuss the pipeline modularization in vllm-omni

File: verl/workers/rollout/replica.py (hunk 7 above)
Comment (SamitHuang) [reply verl/workers/rollout/replica.py]: yes, we can add it after testing with video generation

File: verl/models/diffusion/vllm_omni/qwen_image/qwen_image_transformer.py (hunk 5 above)
Comment (knlnguyen1802) [reply verl/models/diffusion/vllm_omni/qwen_image/qwen_image_transformer.py]: Fixed it

File: examples/vllm_omni/pipeline_qwenimage.py [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+    def __init__(self, *, od_config: OmniDiffusionConfig, prefix: str = ""):
```
Comment (zhtmike) [inline examples/vllm_omni/pipeline_qwenimage.py:40]: suggest to delete `__init__` part, as it is same as with `QwenImagePipeline` after `QwenImageTransformer2DModelFixed` removed.

File: examples/vllm_omni/pipeline_qwenimage.py
Comment (SamitHuang) [inline examples/vllm_omni/pipeline_qwenimage.py:1]: it's better to add a runnable shell script in this example folder

Comment (zhtmike) [reply examples/vllm_omni/pipeline_qwenimage.py:1]: e2e script will be provided once flowgrpo algo is fully integrated. But I suggest to leave a markdown there first

File: examples/vllm_omni/pipeline_qwenimage.py (hunk 9 above)
Comment (knlnguyen1802) [reply examples/vllm_omni/pipeline_qwenimage.py:40]: It's still needed to init the custom scheduler
